Cleaning up GPU executable flatbuffers prior to a larger reworking.

benvanik opened the pull request:
Besides some renaming the major cleanup here is centralizing debug info at the HAL level and adding some utilities for producing/consuming it. Each target may also include additional target-specific debug info (like source maps in WGSL, etc) but the information useful for uniform tracing across targets is now shared.

Future changes will rip apart the GPU flatbuffers to store things per export instead of many arrays and allow for easier addition of export information by modifying the export tables.

This is a breaking change (barely) and is staged into a branch until all changes are complete and a global HAL version bump can happen as part of the dispatch2 work in #18154.

Progress on #18154.

benvanik committed:
Fixing some incorrect TODOs referencing #18154.

benvanik committed:
Fixing ROCM/CUDA BlockSizeDef -> BlockSize.
(tables have Def, structs don't)

benvanik committed:
Renaming Metal kernel_library -> executable.
It's a HAL executable and will contain multiple kernel libraries.

benvanik committed:
Factoring out common debug info from GPU executable flatbuffers.
This also adds source file publishing to all GPU targets. Basic support
for export-specific debug info is added but switching targets to use it
is left to a future change.

benvanik committed:
Renaming rocm executable -> hip executable.

ScottTodd commented on compiler/src/iree/compiler/Dialect/HAL/Utils/ExecutableDebugInfoUtils.h:
Thanks for pulling these functions into a shared helper file!

ScottTodd commented on runtime/src/iree/schemas/executable_debug_info.fbs:
```diff
@@ -0,0 +1,44 @@
+// Copyright 2019 The IREE Authors
```
New-ish file, want to use 2024 instead of 2019?

ScottTodd commented on runtime/src/iree/schemas/hip_executable_def.fbs:
```diff
@@ -6,11 +6,11 @@
-file_extension "rocm";
+// 'HIP Executable'.
+file_identifier "HIPE";
+file_extension "hipe";
```
Could this use just "HIP" instead of "HIPE"? Do the other executables have an "E" suffix?

benvanik replied:
file_identifier has to be 4 characters long (it becomes a uint32_t in the file format)

benvanik replied:
genius idea: `HIP1` so we have a good way to version

ScottTodd reviewed (commented):
Batching breaking changes together SGTM. Thanks for the incremental cleanup.

benvanik committed:
Renaming [spirv|wgsl]_executable_def to [vulkan|webgpu].
I don't expect us to have the same flatbuffers for different HAL
implementations and do expect us to have alternative file formats for
the same HAL implementation.

ScottTodd commented on build_tools/bazel/iree_flatcc.bzl:
```diff
@@ -34,10 +36,10 @@ def iree_flatbuffer_c_library(
 
     native.genrule(
         name = name + "_gen",
-        srcs = srcs,
+        srcs = srcs + includes,
         outs = outs,
         tools = [flatcc],
-        cmd = "$(location %s) %s $(SRCS)" % (flatcc, " ".join(flags)),
+        cmd = "$(location %s) %s %s" % (flatcc, " ".join(flags), " ".join(["$(location {})".format(src) for src in srcs])),
```
Phew, that wasn't so bad. Thanks for keeping Bazel and CMake both working.
